### DISCUSSION TITLE
[ERROR GPL-0305] RePlAce diverged at newStepLength.

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD-flow-scripts#777] [General]
Any tips on what I can look at why the global_place_skip_io.tcl diverges?

There are no congestion heat-maps at this point.

I checked timing: setup, positive slack

```
[NesterovSolve] Iter: 1480 overflow: 2402.21 HPWL: 6560624414
[NesterovSolve] Iter: 1490 overflow: 2402.21 HPWL: 6560624414
[NesterovSolve] Iter: 1500 overflow: 2402.21 HPWL: 6560669009
[ERROR GPL-0305] RePlAce diverged at newStepLength.
Error: global_place_skip_io.tcl, 30 GPL-0305
```

**Comment by maliberty** [comment]
This one is harder to diagnose.  You have a huge amount of overflow which is odd.  Can you include the earlier messages about target density and utilization from the start of placement?

**Reply by oharboe** [reply]
Ah, I found out what is wrong: the macros are outside of the floorplan.

I have found that there's no explicit check and error messages if macros are outside of the floorplan, you get followon errors like the above.

**Reply by oharboe** [reply]
So, if there's an action here, it would be to improve usability for newbies, such as me, by adding an explicit error message if macros are outside of the floorplan and a tip to examine the floorplan.

Macros can end up outside of the floorplan in three cases:

- manual layout of macros
- automatic calculation of the size of the die and core area as macros are disregarded

**Reply by oharboe** [reply]
Ah. Strike that. That macros are outside can cause these problems, but that was not the problem in this case.

**Comment by oharboe** [comment]
This should be reproducable by scaling up this design: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/tree/master/flow/designs/asap7/mock-array-big

There's basically no logic at the top level, just wires between the macros and the outer edges of the design.

More output:

```
[deleted]
[WARNING STA-0164] ./platforms/asap7/lib/asap7sc7p5t_SIMPLE_RVT_TT_nldm_211120.lib.gz line 14838, timing group from output port.
[INFO GPL-0002] DBU: 1000
[INFO GPL-0003] SiteSize: 54 270
[INFO GPL-0004] CoreAreaLxLy: 2052 2160
[INFO GPL-0005] CoreAreaUxUy: 2945268 1489860
[INFO GPL-0006] NumInstances: 597033
[INFO GPL-0007] NumPlaceInstances: 2
[INFO GPL-0008] NumFixedInstances: 290876
[INFO GPL-0009] NumDummyInstances: 306155
[INFO GPL-0010] NumNets: 142261
[INFO GPL-0011] NumPins: 276290
[INFO GPL-0012] DieAreaLxLy: 0 0
[INFO GPL-0013] DieAreaUxUy: 2947270 1492000
[INFO GPL-0014] CoreAreaLxLy: 2052 2160
[INFO GPL-0015] CoreAreaUxUy: 2945268 1489860
[INFO GPL-0016] CoreArea: 4378622443200
[INFO GPL-0017] NonPlaceInstsArea: 2769974465672
[INFO GPL-0018] PlaceInstsArea: 204120
[INFO GPL-0019] Util(%): 0.00
[INFO GPL-0020] StdInstsArea: 204120
[INFO GPL-0021] MacroInstsArea: 0
[INFO GPL-0031] FillerInit: NumGCells: 1576178
[INFO GPL-0032] FillerInit: NumGNets: 142261
[INFO GPL-0033] FillerInit: NumGPins: 276290
[INFO GPL-0023] TargetDensity: 0.10
[INFO GPL-0024] AveragePlaceInstArea: 102060
[INFO GPL-0025] IdealBinArea: 1020600
[INFO GPL-0026] IdealBinCnt: 4290243
[INFO GPL-0027] TotalBinArea: 4378622443200
[INFO GPL-0028] BinCnt: 2048 2048
[INFO GPL-0029] BinSize: 1438 727
[INFO GPL-0030] NumBins: 4194304
[NesterovSolve] Iter: 1 overflow: 2402.21 HPWL: 17377433797
[NesterovSolve] Iter: 10 overflow: 2402.21 HPWL: 17377433797
[this will fail]
```

**Comment by maliberty** [comment]
The overflow at iter 1 == iter 1500 so something is going very wrong.  Is this something you can reproduce on your big array test?

**Reply by oharboe** [reply]
Running test on this one now... https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/779

```[INFO PDN-0001] Inserting grid: top``` will run for a while...

Nope... Doesn't reproduce it.

**Reply by oharboe** [reply]
Trying ludicrous size... 

![image](https://user-images.githubusercontent.com/2798822/213566429-9a2decb7-97fa-4da5-b5e9-6f6f9754cbfa.png)

**Reply by oharboe** [reply]
@maliberty PDN insertion takes "forever", but https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/779 doesn't reproduce the problem in this discussion.

I tried tinkering with the floorplan in my design, and overflow in ```[NesterovSolve] Iter: 1 overflow: 2402.21 HPWL: 17377433797``` got down to about ~900, but was still "stuck" and subsequent iterations didn't reduce it.

**Comment by JiaDYuan** [comment]
Yes, It's very odd that the overflow is so high. I also got an error of "RePlAce diverged at newStepLength". But  here is my final iteration message: [NesterovSolve] Iter: 3380 overflow: 0.138925 HPWL: 51000908303

**Comment by oharboe** [comment]
No longer happens with latest master for my cases.
